### ISSUE TITLE
Feature Request: spaces.Array 

### ISSUE THREAD

**Issue by bionicles** [openai/gym#1274]
I just verified it is not possible to set the default Box space to have a None value in one dimension. This is problematic because often we need to allow inputs of arbitrary lengths -- graphs have different numbers of nodes and edges, songs last variable amounts of time, molecules might be big or small, etc

To solve this I'm testing a new Space called Array which is a clone of Box except without the check for None values. This means we can't sample the values, but really we only use Sample for Actions when we don't have an Agent. We don't usually Sample observations because we can just call Env.reset() and get a real chunk of data...

Array type allows us to specify _partial_ constraints on data -- but then it would be critical for Envs which use this to update the constraints on the Action space after getting data in reset. Otherwise, we can't sample actions when we test the Env.

Another option is to give the Array constructor some default dimensionality and update it later. But then it might fail silently. That's bad. So we'll just move the None checks in Box.__init__ into Array.sample() and if you sample an Array with a None-valued dimension, you get an error

TLDR: 
Box -> All dimensions must be specified -> throws an error if the dimensions are not specified
Array -> Some dimensions may be unspecified -> throws an error during SAMPLING if dimensions are not specified.
 
input welcome!

**Comment by bionicles** [comment]
```
class Array(gym.Space):
    def __init__(self, shape, variance=1., mean=0., high=None, low=None, dtype=np.float32):
        self.shape = shape
        self.dtype = dtype
        # sampling
        self.variance = variance
        self.mean = mean
        # constraints
        self.high = high
        self.low = low

    def sample(self):
        if self.shape is not None and None in self.shape:
            raise AttributeError("cannot sample arrays with unspecified dimensionality")
        else:
            sample = np.random.normal(self.mean, self.variance, self.shape)
            if self.low is not None and self.high is not None:
                sample = np.clip(sample, self.low, self.high)
            return sample

    def contains(self, x):
        if x.shape is self.shape:
            if all(self.low < x < self.high):
                return True
        return False
```

**Comment by bionicles** [comment]
need to handle integer sampling case ... but it's a start

**Comment by bionicles** [comment]
one issue with space.Dict is it doesn't support item assignment: this means we have to create a new space during reset instead of updating the original spaces.Dict object.

how do you make spaces.Dict easier to use?

it would be great if the Dict space were similar to AttrDict ---> then we could set and get attribute values from the dict easily like this:

`self.action_space = spaces.Dict( { "potentials": spaces.Array(shape=(None, 3)) } `

... later in reset we get num_atoms so we can set the shape of the potentials array in the action space:

`self.action_space.potentials.shape[0]  == num_atoms`

right now we have to create a whole new object like this:
`self.action_space = spaces.Dict( { "potentials": spaces.Array(shape=(num_atoms, 3)) } `

**Comment by Eriz11** [comment]
@bionicles ¿Have you finally implemented this? I'm also interested in this specific case. 

In my case, I have an observation_space within a spaces.Dict. However, one of the input arrays is of the same length every time (say shape=(10,3)) but the other input array can be of an arbitrary length (i.e. sometimes shape=(15,2), other time shape=(20,2) etc.). 

Any improvement in that would be awesome.

**Comment by bionicles** [comment]
Yeah  it works, I’ll check to see if I tinkered more but it’s basically a fork of box with the ability to set None dimensionality

**Comment by Eriz11** [comment]
That's nice. Many thanks for the work. 

¿Could you illustrate how to integrate the new space into the framework and use it like the other spaces?

Also, and maybe it is a silly question because on my early deepen into RL, ¿Does having variable length observations function properly within the agents? For example, in some DL algorithms we must fill with zeros when the inputs are not of the same shape. 

Thanks in advance for the guidance,

**Comment by christopherhesse** [comment]
Seems interesting, if you end up packaging that as a library, might be nice to link from this issue.  I don't think we're going to change the built-in gym spaces at this time though.
